trieste/models/keras/sampler.py

```diff
@@ -0,0 +1,116 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module is the home of the sampling functionality required by some
+of the Trieste's Keras model wrappers.
+"""
+
+from __future__ import annotations
+
+import tensorflow as tf
+
+from ...types import TensorType
+from ..interfaces import (
+    EnsembleModel,
+    TrajectoryFunction,
+    TrajectoryFunctionClass,
+    TrajectorySampler,
+)
+
+
+class EnsembleTrajectorySampler(TrajectorySampler[EnsembleModel]):
+    """
+    This class builds functions that approximate a trajectory by randomly choosing a network from
+    the ensemble and using its predicted means as a trajectory.
+    """
+
+    def __init__(self, model: EnsembleModel):
+        """
+        :param model: The ensemble model to sample from.
+        """
+        if not isinstance(model, EnsembleModel):
+            raise NotImplementedError(
+                f"EnsembleTrajectorySampler only works with EnsembleModel models, that support "
+                f"ensemble_size, sample_index, predict_ensemble and sample_ensemble methods; "
+                f"received {model.__repr__()}"
+            )
+
+        super().__init__(model)
+
+        self._model = model
+
+    def __repr__(self) -> str:
+        """"""
+        return f"{self.__class__.__name__}({self._model!r}"
+
+    def get_trajectory(self, negate: bool = False) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) by randomly choosing a network from
+        the ensemble and using its predicted means as a trajectory.
+
+        :param negate: Return the negative of the trajectory.
+        :return: A trajectory function representing an approximate trajectory
+            from the model, taking an input of shape `[N, 1, D]` and returning shape `[N, 1]`.
+        """
+        return ensemble_trajectory(self._model, negate)
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
+        """
+        Efficiently resample a :const:`TrajectoryFunction` in-place to avoid function retracing
+        with every new sample.
+
+        :param trajectory: The trajectory function to be resampled.
+        :return: The new resampled trajectory function.
+        """
+        tf.debugging.Assert(isinstance(trajectory, ensemble_trajectory), [])
+        trajectory.resample()  # type: ignore
+        return trajectory
+
+
+class ensemble_trajectory(TrajectoryFunctionClass):
+    """
+    Generate an approximate function draw (trajectory) by randomly choosing a network from
```

Comment:
potential limitation here is the number of networks in the ensemble, thats giving us the number of approximations available

an alternative might be to sample from the network's output distribution rather than use the mean, it would be more costly, but not as much as sampling from a GP, so perhaps good enough? or leave it as an option?

@henrymoss and @vpicheny any thoughts?

Reply:
I think that we should stick to the simple solution, which would be to check that you have more models than query points and throw an error if not?

The other case seems interesting but much more on the research side: at least for now, we have no evidence that this is a good idea and it's not that natural, so maybe we don't want to support it in the short term?

Reply:
I think you mean more than batch size not query points?
not sure how we can check this here, perhaps if I allow batch variant (I restrict it below but actually nothing prevents us to return a batch directly here...) I can test that

good enough for the second point, I'll test that one first...

Reply:
So, I have a draft where we pass a` num_samples` arg to `get_trajectory` as this is important when using our parallel acq optimiser to efficiently optimize multiple trajectories in parallel.  I'm not ready to merge this one yet. But feel free to add in this argument (making the default value=1 and making my RFF sampler raise an error for now when >1) ?

Reply:
ah, great to hear that, I'll merge this in now and revise it after you are done with yours